pkg/server/commands/list_objects.go

```diff
@@ -90,86 +89,86 @@ func (q *ListObjectsQuery) evaluate(
-		userObj, userRel := tuple.SplitObjectRelation(req.GetUser())
+	if tuple.IsTypedWildcard(userObj) {
+		sourceUserRef = &UserRefTypedWildcard{Type: tuple.GetType(userObj)}
 
-		userObjType, userObjID := tuple.SplitObject(userObj)
+	}
 
-		var targetUserRef isUserRef
-		targetUserRef = &UserRefObject{
-			Object: &openfgapb.Object{
-				Type: userObjType,
-				Id:   userObjID,
+	if userRel != "" {
+		sourceUserRef = &UserRefObjectRelation{
+			ObjectRelation: &openfgapb.ObjectRelation{
+				Object:   userObj,
+				Relation: userRel,
 			},
 		}
+	}
 
-		if tuple.IsTypedWildcard(userObj) {
-			targetUserRef = &UserRefTypedWildcard{Type: tuple.GetType(userObj)}
-		}
+	connectedObjectsResChan := make(chan *ConnectedObjectsResult, 1)
 
-		if userRel != "" {
-			targetUserRef = &UserRefObjectRelation{
-				ObjectRelation: &openfgapb.ObjectRelation{
-					Object:   userObj,
-					Relation: userRel,
-				},
-			}
+	var objectsFound = new(uint32)
+
+	go func() {
+		err = q.ConnectedObjects(ctx, &ConnectedObjectsRequest{
+			StoreID:          req.GetStoreId(),
+			Typesystem:       typesys,
+			ObjectType:       targetObjectType,
+			Relation:         targetRelation,
+			User:             sourceUserRef,
+			ContextualTuples: req.GetContextualTuples().GetTupleKeys(),
+		}, connectedObjectsResChan)
+		if err != nil {
+			errChan <- err
+			close(errChan)
 		}
 
-		handler = func() {
-			span.SetAttributes(attribute.Bool(listObjectsOptimizedKey, true))
+		close(connectedObjectsResChan)
+	}()
 
-			connectObjCmd := &ConnectedObjectsCommand{
-				Datastore:        q.Datastore,
-				ResolveNodeLimit: q.ResolveNodeLimit,
-				Limit:            maxResults,
+	go func() {
+
+		for res := range connectedObjectsResChan {
+
+			if res.ResultStatus == NoFurtherEvalStatus && atomic.AddUint32(objectsFound, 1) <= maxResults {
+				resultsChan <- res.Object
+				continue
 			}
 
-			err = connectObjCmd.StreamedConnectedObjects(ctx, &ConnectedObjectsRequest{
-				StoreID:          req.GetStoreId(),
-				Typesystem:       typesys,
-				ObjectType:       targetObjectType,
-				Relation:         targetRelation,
-				User:             targetUserRef,
-				ContextualTuples: req.GetContextualTuples().GetTupleKeys(),
-			}, resultsChan)
+			// todo(optimization): fire this Check resolution off concurrently and continue on to the next object in the
+			// result channel
+			resp, err := q.CheckResolver.ResolveCheck(ctx, &graph.ResolveCheckRequest{
+				StoreID:              req.GetStoreId(),
+				AuthorizationModelID: req.GetAuthorizationModelId(),
+				TupleKey:             tuple.NewTupleKey(res.Object, req.GetRelation(), req.GetUser()),
+				ContextualTuples:     req.GetContextualTuples().GetTupleKeys(),
+				ResolutionMetadata: &graph.ResolutionMetadata{
+					Depth: q.ResolveNodeLimit,
+				},
+			})
 			if err != nil {
 				errChan <- err
 				close(errChan)
-				return
+				return // is this the right thing here or should we try to find all of them we can up to some limit?
```

Comment:
That seems like broken behavior in our current implementation. We should debate that as a maintainer team.

I think the behavior should match similar expectations with Check. That is, if we see at least 1 error and cannot fulfill the max results limit with the other evaluations, then we return an error. This way we do our best to satisfy the request with a complete response but in the event we cannot then we return an error. This is similar to our Check behavior where we hold off on returning an error until we've evaluated all paths that could lead to a result without an error response.

Reply:
> if we see at least 1 error and cannot fulfill the max results limit with the other evaluations, then we return an error. 

This will be fine for StreamedListObjects, where we return results as we get them, but for the regular endpoint, clients will get no results at all if there's one error somewhere.

Reply:
Let's talk about that behavior as a maintainer team. 

I think this current behavior is a safer option than potentially returning incomplete results. If you can fill the max_results limit despite an error happening, then I think it's ok to swallow the error. However, if you cannot fill the max_results limit and an error happened with at least one of the evaluations, then I don't think the error should be swallowed.

Reply:
As per our discussion offline as a maintainer team, we are going to favor returning an error fast if one occurs, because doing so gives us less of a chance to return partial results that could be misleading.